e2e-test/src/api.rs

```diff
@@ -0,0 +1,310 @@
+// This file is part of Astarte.
+//
+// Copyright 2024 SECO Mind Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::HashMap, fmt::Debug, str::FromStr};
+
+use astarte_device_sdk::types::AstarteType;
+use chrono::{DateTime, Utc};
+use color_eyre::{owo_colors::OwoColorize, Section, SectionExt};
+use eyre::{ensure, eyre};
+use reqwest::{Response, Url};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_json::Value;
+use tracing::{debug, instrument, trace};
+
+use crate::utils::{base64_decode, base64_encode, read_env, Timestamp};
+
+#[derive(Clone)]
+pub struct Api {
+    ///  Base url plus the realm and device id
+    url: Url,
+    token: String,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct ApiData<T> {
+    pub data: T,
+}
+
+impl<T> ApiData<T> {
+    fn new(data: T) -> Self
+    where
+        T: Serialize,
+    {
+        Self { data }
+    }
+}
+
+#[derive(Deserialize)]
+struct WithTimestamp<T> {
+    #[serde(flatten)]
+    value: T,
+    #[serde(rename = "timestamp")]
+    _timestamp: Option<DateTime<Utc>>,
+}
+
+impl Debug for Api {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Api")
+            .field("url", &self.url)
+            .finish_non_exhaustive()
+    }
+}
+
+async fn check_response(url: &str, res: Response) -> eyre::Result<Response> {
+    let status = res.status();
+    if status.is_client_error() || status.is_server_error() {
+        let body: Value = res.json().await?;
+
+        Err(eyre!("HTTP status error ({status}) for url {url}")
+            .section(format!("The response body is:\n{body:#}").header("Response:".cyan())))
+    } else {
+        Ok(res)
+    }
+}
+
+impl Api {
+    fn url(api_url: &str, realm: &str, device_id: &str) -> eyre::Result<Url> {
+        let url = Url::parse(&format!("{api_url}/v1/{realm}/devices/{device_id}"))?;
+
+        Ok(url)
```

Comment:
Couldn't you simply retun `Url::parse(...)` (without the let assignment)?

Reply:
I would still have to `.map_err(Into::into)` instead of the try `?`